### ISSUE TITLE
Frequent crashes when using pixel classifier 0.2.0 m8

### ISSUE THREAD

**Issue by MakarowMax** [qupath/qupath#394]
## Bug report

**Describe the bug**
QuPath crashes very frequently when I try to use the pixel classifier tool. The program just closes without any further warning or error message. Sometimes an error report is generated an saved. 

**To Reproduce**

1. Go to 'Classify'
2. Click on 'Train pixel classifier (experimental)'
3. Settings: 

- Classifier: Random trees, Parameters: Maximum tree depth: 20, Minimum sample count: 10, 
- Resolution: Full (0.3 µm/px), 
- Features: Scales: 0.5, 1.0, 2.0, 4.0, 8.0, Features: Gaussian, Gradient magnitude, 
- Advanced options: Feature normalization: Mean & variance, Boundary strategy: Classify as Stroma, Boundary thickness: 2 pixels

The crash happens when I try to change the settings above, while 'Live prediction' runs.

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x0000000000cb5b26, pid=9416, tid=7376
#
# JRE version: OpenJDK Runtime Environment (13.0.1+9) (build 13.0.1+9)
# Java VM: OpenJDK 64-Bit Server VM (13.0.1+9, mixed mode, tiered, compressed oops, g1 gc, windows-amd64)
# Problematic frame:
# C  0x0000000000cb5b26
#
# No core dump will be written. Minidumps are not enabled by default on client versions of Windows
#
# If you would like to submit a bug report, please visit:
#   https://github.com/AdoptOpenJDK/openjdk-build/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  S U M M A R Y ------------

Command Line: -Djava.library.path=C:\Users\ag5-35\Downloads\QuPath-0.2.0-m8-Windows\QuPath-0.2.0-m8\app;C:\Users\ag5-35\Downloads\QuPath-0.2.0-m8-Windows\QuPath-0.2.0-m8 -Djava.launcher.path=C:\Users\ag5-35\Downloads\QuPath-0.2.0-m8-Windows\QuPath-0.2.0-m8 -Xmx8192M -XX:MaxRAMPercentage=50 qupath.QuPath

Host: Intel(R) Core(TM) i5-9400 CPU @ 2.90GHz, 6 cores, 15G,  Windows 10 , 64 bit Build 18362 (10.0.18362.329)
Time: Tue Jan 28 18:35:40 2020 W. Europe Standard Time elapsed time: 148 seconds (0d 0h 2m 28s)

---------------  T H R E A D  ---------------

Current thread (0x000001d550e5e800):  JavaThread "classifier-overlay6" daemon [_thread_in_native, id=7376, stack(0x0000002279b00000,0x0000002279c00000)]

Stack: [0x0000002279b00000,0x0000002279c00000],  sp=0x0000002279bf9e20,  free space=999k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  0x0000000000cb5b26

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  org.bytedeco.opencv.opencv_ml.LogisticRegression.predict(Lorg/bytedeco/opencv/opencv_core/Mat;Lorg/bytedeco/opencv/opencv_core/Mat;I)F+0
j  qupath.opencv.ml.OpenCVClassifiers$AbstractOpenCVClassifierML.predictWithLock(Lorg/bytedeco/opencv/opencv_core/Mat;Lorg/bytedeco/opencv/opencv_core/Mat;Lorg/bytedeco/opencv/opencv_core/Mat;)V+11
j  qupath.opencv.ml.OpenCVClassifiers$AbstractOpenCVClassifierML.predict(Lorg/bytedeco/opencv/opencv_core/Mat;Lorg/bytedeco/opencv/opencv_core/Mat;Lorg/bytedeco/opencv/opencv_core/Mat;)V+14
j  qupath.opencv.ml.OpenCVClassifiers$LogisticRegressionClassifier.predict(Lorg/bytedeco/opencv/opencv_core/Mat;Lorg/bytedeco/opencv/opencv_core/Mat;Lorg/bytedeco/opencv/opencv_core/Mat;)V+4
j  qupath.opencv.ml.pixel.OpenCVPixelClassifier.applyClassification(Lqupath/lib/images/ImageData;Lqupath/lib/regions/RegionRequest;)Ljava/awt/image/BufferedImage;+235
j  qupath.lib.classifiers.pixel.PixelClassificationImageServer.readTile(Lqupath/lib/images/servers/TileRequest;)Ljava/awt/image/BufferedImage;+84
J 14172 c1 qupath.lib.images.servers.AbstractTileableImageServer.getTile(Lqupath/lib/images/servers/TileRequest;)Ljava/awt/image/BufferedImage; (133 bytes) @ 0x000001d52232b9dc [0x000001d52232b3e0+0x00000000000005fc]
J 14898 c1 qupath.lib.images.servers.AbstractTileableImageServer.readBufferedImage(Lqupath/lib/regions/RegionRequest;)Ljava/awt/image/BufferedImage; (1110 bytes) @ 0x000001d522ceb264 [0x000001d522ce2920+0x0000000000008944]
j  qupath.lib.images.servers.AbstractTileableImageServer.readBufferedImage(Lqupath/lib/regions/RegionRequest;)Ljava/lang/Object;+2
j  qupath.lib.gui.ml.PixelClassificationOverlay.lambda$requestTile$2(Lqupath/lib/images/servers/TileRequest;Lqupath/lib/images/servers/ImageServer;)V+32
j  qupath.lib.gui.ml.PixelClassificationOverlay$$Lambda$1631.run()V+12
J 12931 c1 java.util.concurrent.Executors$RunnableAdapter.call()Ljava/lang/Object; java.base@13.0.1 (14 bytes) @ 0x000001d521d305d4 [0x000001d521d304c0+0x0000000000000114]
J 14612 c1 java.util.concurrent.FutureTask.run()V java.base@13.0.1 (123 bytes) @ 0x000001d52225e6fc [0x000001d52225e020+0x00000000000006dc]
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+92 java.base@13.0.1
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5 java.base@13.0.1
j  java.lang.Thread.run()V+11 java.base@13.0.1
v  ~StubRoutines::call_stub

siginfo: EXCEPTION_ACCESS_VIOLATION (0xc0000005), reading address 0xffffffffffffffff

...
```
The crash doesn't happen consistently.

**Expected behavior**
I want to see a live prediction of the classifier and save it. Then I want to apply the classifier to my images.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - QuPath Version 0.2.0 m8

**Comment by petebankhead** [comment]
Thanks @MakarowMax, I haven't seen this error before but I am able to replicate it when I chance to the *Logistic Regression* classifier. I can change to *Artificial Neural Network* without problems. Is this also true for you?

It's curious because I am fairly sure I have used *Logistic Regression* without errors in the past; I'm not sure if it depends upon operating system or if any recent updates might be causing the trouble. In any case, the error occurs deep within OpenCV - and unfortunately there is no possibility to capture the exception within QuPath before the Java Virtual Machine shuts down.

In the past, the only solution I could find to OpenCV errors like this has been to avoid calling the crashing code. In this case, that would mean removing the *Logistic Regression* option. I haven't found it to be terribly useful (even when it worked) so I don't think this would be a big loss; I personally prefer *Artificial Neural Network* (and sometimes *Random Trees*). *K Nearest Neighbor* really only makes sense when training using point annotations (when given a large number of samples it is too slow).

What do you think? Does it happen to you with other classifier types as well?

**Comment by MakarowMax** [comment]
Thank you for your quick response, Pete. Today morning I tried to provoke this error with different classifiers. The error occured with 'Random trees', 'Artificial neural network' and 'Logistic regression'. I think the more settings I change, the more likely the error will occur - but I couldn't figure out if a single setting is the reason.

With my usual approach I train the pixel classifier on a project training image. If I could successfully save the classifier and load it for another image in the project, I get the following error messages for ANN and KNN:

ANN classifier:
![err_ann](https://user-images.githubusercontent.com/59967282/73450688-34d01a00-4366-11ea-9c8f-7e08251c6261.png)

KNN classifier:
![err_knn](https://user-images.githubusercontent.com/59967282/73450694-37cb0a80-4366-11ea-98f8-17df5cefcf52.png)

**Comment by petebankhead** [comment]
I think the saving/reloading errors are completely different from the complete crashes. I'm a bit lost regarding which occurs with which classifiers and when, and I can't replicate most of what you report (except for the logistic regression thing).

I also notice that you're running your classifiers at a very high resolution, and also with the parameters you're using the model could become rather complex. I haven't really had a chance to explore the limits of the pixel classifier, but it may well be that it really requires simpler models in its current form (the really complex stuff will likely need deep learning... and therefore quite a different approach).

With that in mind

* Do you experience the same errors when using the default parameters?
* Do you experience the same errors when using simpler models (e.g. less training data, applied at a lower resolution)?
* If you try things on another computer, do you see exactly the same behavior?

**Comment by petebankhead** [comment]
@MakarowMax are you able to share a classifier file that will not open (it should be in JSON format, stored within the project)?

My guess is that your classifiers are particularly large and complex as this could be part of the issue, but I have no idea if this is correct and have not been able to replicate most of the issues.

**Comment by MakarowMax** [comment]
@petebankhead sorry for the late answer, but I finally had time to continue with this project. No I never experienced any errors when I used default parameters. Simpler models with lower resolution greatly reduce the error rate! So I guess you are right, that my classifiers were too complex and that this is part of the problem. I tried my approach on four different computers and I've had the same errors every time, when I started the pixel classifier on full resolution. 


I've added one of my JSON files below.

[2020-02-21 Pixel Model.zip](https://github.com/qupath/qupath/files/4235332/2020-02-21.Pixel.Model.zip)

**Comment by petebankhead** [comment]
The classifier is indeed extremely large, although it loads for me on both Mac and Windows using v0.2.0-m9. Which exact problem do you see with this file?

**Comment by petebankhead** [comment]
v0.2.0-m10 is now available, which contains an updated version of OpenCV that I think/hope should have fixed the issue. But since I couldn't reproduce it in the first place, I'm not certain.

I'll close this issue, but if you find the problem remains - or you can confirm it's resolved in m10 - please do let me know.
